### PR TITLE
Crunchy Postgres Exporter

### PR DESCRIPTION
**Pull request by tjmoore4**
**Checklist:**

 
 - [x] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [x] Have you updated or added documentation for the change, as applicable?
 - [x] Have you tested your changes on all related environments with successful results, as applicable?

**Type of Changes:**

 
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] New feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change)

**What is the current behavior? (link to any open issues here)**
The project does not contain the Crunchy Postgres Exporter
container.

**What is the new behavior (if this is a feature change)?**
This commit moves the 'crunchy-collect' container,
now called 'crunchy-postgres-exporter' from the
"Crunchy Containers" repository to the "Postgres
Operator" repository. This commit then provides all
the required build and code changes for this container to
work as expected.

This commit also adds the relevant documentation for
the Crunchy Postgres Exporter container.

**Other information**:
[ch8734]

### CODE REVIEW

File: bin/crunchy-exporter/start.sh [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+set_exporter_pg_credentials() {
+
+    if [[ -d "/exporter-pguser" ]]
+    then
+	    echo_info "Setting credentials for exporter PG user using file system"
+
+	    EXPORTER_PG_USER=$(cat /exporter-pguser/username)
+	    export EXPORTER_PG_USER
+
+        EXPORTER_PG_PASSWORD=$(cat /exporter-pguser/password)
```
Comment (andrewlecuyer) [inline bin/crunchy-exporter/start.sh:83]: Some indentation off here.

File: bin/crunchy-exporter/exporter_uid_daemon.sh [hunk 2]
```diff
@@ -0,0 +1,21 @@
+#!/bin/sh
```
Comment (andrewlecuyer) [inline bin/crunchy-exporter/exporter_uid_daemon.sh:1]: Can we use the existing `uid_daemon.sh` that is already under `bin`?

File: centos7/Dockerfile.crunchy-postgres-exporter.centos7 [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+RUN chmod g=u /etc/passwd
+
+# The VOLUME directive must appear after all RUN directives to ensure the proper
+# volume permissions are applied when building the image
+VOLUME ["/conf"]
+
+ENTRYPOINT ["/opt/cpm/bin/exporter_uid_daemon.sh"]
+
+USER 26
```
Comment (andrewlecuyer) [inline centos7/Dockerfile.crunchy-postgres-exporter.centos7:40]: I see that that the entrypoint calls a script to set the UID for the daemon user, but the user has been updated to `26` here (i.e. the postgres user).  Should this be `USER 2`?

File: conf/crunchy-exporter/group.template [hunk 4]
```diff
@@ -0,0 +1,33 @@
+root:x:0:
```
Comment (andrewlecuyer) [inline conf/crunchy-exporter/group.template:1]: Do we need this file?  I know `common_lib.sh` has the `ose_hack` function that references that file, but don't think we'll be using that here.

Can we therefore leave out this file and remove that unneeded function?

File: docs/content/advanced/crunchy-postgres-exporter.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+## Packages
+
+The crunchy-postgres-exporter Docker image contains the following packages (versions vary depending on PostgreSQL version):
+
+* PostgreSQL (12.3, 11.8, 10.13, 9.6.18 and 9.5.22)
+* CentOS7 - publicly available
+* UBI7 - customers only
+* [PostgreSQL Exporter](https://github.com/wrouesnel/postgres_exporter)
+
+## Environment Variables
```
Comment (andrewlecuyer) [inline docs/content/advanced/crunchy-postgres-exporter.md:31]: Should or will the user ever be directly modifying the env vars for this container themselves?

We don't typically document Operator-specific container env vars since the user influences config via `pgcluster` specs (and therefore CLI flags, etc.) and not directly via env vars, so wanted to clarify the motivation here.

File: bin/crunchy-exporter/start.sh (hunk 1 above)
Comment (tjmoore4) [reply bin/crunchy-exporter/start.sh:83]: yep, some tab/space inconsistencies. Updated, thanks.

File: centos7/Dockerfile.crunchy-postgres-exporter.centos7 (hunk 3 above)
Comment (tjmoore4) [reply centos7/Dockerfile.crunchy-postgres-exporter.centos7:40]: Yes, in the current design, that should now be 2. Updated.

File: conf/crunchy-exporter/group.template (hunk 4 above)
Comment (tjmoore4) [reply conf/crunchy-exporter/group.template:1]: Ok, thanks for that info. I've removed the template files and initial testing confirms the container is working as expected.

File: bin/crunchy-exporter/exporter_uid_daemon.sh (hunk 2 above)
Comment (tjmoore4) [reply bin/crunchy-exporter/exporter_uid_daemon.sh:1]: Sure, on review, the differences between the files seem to be inconsequential for this implementation. Updated, thanks.

File: docs/content/advanced/crunchy-postgres-exporter.md (hunk 5 above)
Comment (tjmoore4) [reply docs/content/advanced/crunchy-postgres-exporter.md:31]: The updated documentation provides the info needed to connect the various metrics available through the container, so the extra details may be more useful here than for other containers in the project.

Since this was posted, I added some more details to hopefully provide better context. Let me know if there might be anything else that should be added or deleted, and I'm happy to tweak what's here.

File: bin/crunchy-exporter/common_lib.sh [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+function pgisready() {
+    export PGROOT=$(find /usr/ -type d -name 'pgsql-*')
+    local dbname=${1?}
+    local dbhost=${2?}
+    local dbport=${3?}
+    local dbuser=${4?}
+    local max_attempts=${5:-5}
+    local timeout=${6:-2}
+    test_server ${dbname?} ${dbhost?} ${dbport?} ${dbuser?} ${max_attempts?} ${timeout?}
+    test_query ${dbname?} ${dbhost?} ${dbport?} ${dbuser?} ${max_attempts?} ${timeout?}
```
Comment (jkatz) [inline bin/crunchy-exporter/common_lib.sh:88]: I'm not sure why we need `test_query`. If we're able to connect, we're able to query for the purposes of the exporter, as we're only making read-only queries.

File: bin/crunchy-exporter/common_lib.sh [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+        attempt=$(( attempt + 1 ))
+        timeout=$(( timeout * 2 ))
+    done
+
+    if [[ ${error?} == 'true' ]]
+    then
+        echo_err "Could not run query against PostgreSQL: Host=${dbhost?}:${dbport?} DB=${dbname?} User=${dbuser?}"
+        exit 1
+    fi
+}
```
Comment (jkatz) [inline bin/crunchy-exporter/common_lib.sh:160]: I think we can remove this.

File: bin/crunchy-exporter/start.sh [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+
+# Check that postgres is accepting connections.
+echo_info "Waiting for PostgreSQL to be ready.."
+while true; do
+    ${PG_DIR?}/bin/pg_isready -d ${DATA_SOURCE_NAME}
+    if [ $? -eq 0 ]; then
+        break
+    fi
+    sleep 2
+done
```
Comment (jkatz) [inline bin/crunchy-exporter/start.sh:107]: I see this here, and our `pgisready` func defined in the common file. I think we should either use this or that. I'm fine with this one.

File: bin/pre-pull-crunchy-containers.sh [hunk 9]
```diff
@@ -13,7 +13,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-for CNAME in crunchy-postgres crunchy-collect crunchy-pgbadger crunchy-pgbouncer crunchy-grafana crunchy-prometheus
+for CNAME in crunchy-postgres crunchy-pgbadger crunchy-pgbouncer crunchy-grafana crunchy-prometheus
```
Comment (jkatz) [inline bin/pre-pull-crunchy-containers.sh:16]: I believe we can take out `crunchy-grafana` and `crunchy-prometheus` here as well, though in a separate commit.

File: conf/postgres-operator/cluster-deployment.json [hunk 10]
```diff
@@ -0,0 +1,360 @@
+{
```
Comment (jkatz) [inline conf/postgres-operator/cluster-deployment.json:1]: I think this file needs to be removed?

File: conf/postgres-operator/exporter.json [hunk 11]
```diff
@@ -0,0 +1,58 @@
+,{
```
Comment (jkatz) [inline conf/postgres-operator/exporter.json:1]: I think this file needs to be moved to the consolidated installation files?

File: docs/content/architecture/postgres-operator-containers-overview.md [hunk 12]
```diff
@@ -36,9 +36,9 @@ The PostgreSQL Operator orchestrates a series of PostgreSQL and PostgreSQL relat
 
 ### Metrics and Monitoring
 
-* **Metrics Collection** (crunchy-collect). The crunchy-collect container provides real time metrics about the PostgreSQL database via an API. These metrics are scraped and stored by a Prometheus time-series database and are then graphed and visualized through the open source data visualizer Grafana.  
+* **Metrics Collection** (crunchy-postgres-exporter). The crunchy-postgres-exporter container provides real time metrics about the PostgreSQL database via an API. These metrics are scraped and stored by a Prometheus time-series database and are then graphed and visualized through the open source data visualizer Grafana.  
 
-* **Grafana** (crunchy-grafana).  Visual dashboards are created from the collected and stored data that crunchy-collect and crunchy-prometheus provide for the crunchy-grafana container, which hosts an open source web-based graphing dashboard called Grafana.
+* **Grafana** (crunchy-grafana).  Visual dashboards are created from the collected and stored data that crunchy-postgres-exporter and crunchy-prometheus provide for the crunchy-grafana container, which hosts an open source web-based graphing dashboard called Grafana.
```
Comment (cbandy) [inline docs/content/architecture/postgres-operator-containers-overview.md:41]: This whole sentence seems awkward. The one in #1757 looks good to me and doesn't include container names.

Perhaps revert this change so that it doesn't conflict during merge?

File: installers/olm/postgresoperator.csv.images.yaml [hunk 13]
```diff
@@ -13,15 +13,15 @@
...
-- { name: RELATED_IMAGE_CRUNCHY_PGBOUNCER,        value: '${CCP_IMAGE_PREFIX}/crunchy-pgbouncer:${CCP_IMAGE_TAG}' }
-- { name: RELATED_IMAGE_CRUNCHY_PGDUMP,           value: '${CCP_IMAGE_PREFIX}/crunchy-pgdump:${CCP_IMAGE_TAG}' }
-- { name: RELATED_IMAGE_CRUNCHY_PGRESTORE,        value: '${CCP_IMAGE_PREFIX}/crunchy-pgrestore:${CCP_IMAGE_TAG}' }
-- { name: RELATED_IMAGE_CRUNCHY_POSTGRES_HA,      value: '${CCP_IMAGE_PREFIX}/crunchy-postgres-ha:${CCP_IMAGE_TAG}' }
-- { name: RELATED_IMAGE_CRUNCHY_POSTGRES_GIS_HA,  value: '${CCP_IMAGE_PREFIX}/crunchy-postgres-gis-ha:${CCP_POSTGIS_IMAGE_TAG}' }
-- { name: RELATED_IMAGE_CRUNCHY_PROMETHEUS,       value: '${CCP_IMAGE_PREFIX}/crunchy-prometheus:${CCP_IMAGE_TAG}' }
+- { name: RELATED_IMAGE_CRUNCHY_ADMIN,               value: '${CCP_IMAGE_PREFIX}/crunchy-admin:${CCP_IMAGE_TAG}' }
+- { name: RELATED_IMAGE_CRUNCHY_BACKREST_RESTORE,    value: '${CCP_IMAGE_PREFIX}/crunchy-backrest-restore:${CCP_IMAGE_TAG}' }
+- { name: RELATED_IMAGE_CRUNCHY_POSTGRES_EXPORTER,   value: '${CCP_IMAGE_PREFIX}/crunchy-postgres-exporter:${CCP_IMAGE_TAG}' }
```
Comment (cbandy) [inline installers/olm/postgresoperator.csv.images.yaml:18]: Images built in this project are tagged using `PGO_IMAGE_PREFIX`.

File: internal/operator/clusterutilities.go [hunk 14]
```diff
@@ -400,14 +402,16 @@ func CreatePGHAConfigMap(clientset kubernetes.Interface, cluster *crv1.Pgcluster
 	return nil
 }
 
-// sets the proper collect secret in the deployment spec if collect is enabled
-func GetCollectVolume(clientset kubernetes.Interface, cl *crv1.Pgcluster, namespace string) string {
-	if cl.Spec.UserLabels[config.LABEL_COLLECT] == "true" {
-		return "\"secret\": { \"secretName\": \"" + cl.Spec.CollectSecretName + "\" }"
+/*
+// sets the proper exporter secret in the deployment spec if exporter is enabled
+func GetExporterVolume(clientset kubernetes.Interface, cl *crv1.Pgcluster, namespace string) string {
```
Comment (cbandy) [inline internal/operator/clusterutilities.go:407]: Please delete this instead of leaving it commented out.

File: internal/operator/cluster/cluster.go [hunk 15]
```diff
@@ -59,9 +59,9 @@ const ReplicaSuffix = "-replica"
 
 // contstants defining the names of the various sidecar containers
 const (
-	collectCCPImage    = "crunchy-collect"
-	pgBadgerCCPImage   = "crunchy-pgbadger"
-	crunchyadmCCPImage = "crunchy-admin"
+	postgresexporterPGOImage = "crunchy-postgres-exporter"
```
Comment (cbandy) [inline internal/operator/cluster/cluster.go:62]: I don't see this constant being used. Can we remove it?

File: conf/postgres-operator/cluster-deployment.json (hunk 10 above)
Comment (tjmoore4) [reply conf/postgres-operator/cluster-deployment.json:1]: Agreed, it looks like it was added back during the most recent rebase. Removed, thanks.

File: conf/postgres-operator/exporter.json (hunk 11 above)
Comment (tjmoore4) [reply conf/postgres-operator/exporter.json:1]: Yes, similar to the issue above. Updated.

File: bin/crunchy-exporter/common_lib.sh (hunk 7 above)
Comment (tjmoore4) [reply bin/crunchy-exporter/common_lib.sh:160]: Sure thing, updated.

File: bin/crunchy-exporter/common_lib.sh (hunk 6 above)
Comment (tjmoore4) [reply bin/crunchy-exporter/common_lib.sh:88]: :+1: removed, thanks.

File: bin/crunchy-exporter/start.sh (hunk 8 above)
Comment (tjmoore4) [reply bin/crunchy-exporter/start.sh:107]: Works for me. removed `pgisready()` from the common_lib.sh

File: installers/olm/postgresoperator.csv.images.yaml (hunk 13 above)
Comment (tjmoore4) [reply installers/olm/postgresoperator.csv.images.yaml:18]: Updated, thanks.

File: internal/operator/clusterutilities.go (hunk 14 above)
Comment (tjmoore4) [reply internal/operator/clusterutilities.go:407]: Got it, must have missed that in cleanup. Thanks!

File: internal/operator/cluster/cluster.go (hunk 15 above)
Comment (tjmoore4) [reply internal/operator/cluster/cluster.go:62]: Yes, those all three constants are no longer in use, so I removed them.

File: docs/content/architecture/postgres-operator-containers-overview.md (hunk 12 above)
Comment (tjmoore4) [reply docs/content/architecture/postgres-operator-containers-overview.md:41]: No problem. I haven't actually rebased to include 1757 yet, so I can get that when I do.

File: bin/crunchy-exporter/common_lib.sh (hunk 7 above)
Comment (tjmoore4) [reply bin/crunchy-exporter/common_lib.sh:160]: On further review, there were quite a few functions that could be cleared out with the new changes, so I've removed those as well.

File: bin/pre-pull-crunchy-containers.sh (hunk 9 above)
Comment (tjmoore4) [reply bin/pre-pull-crunchy-containers.sh]: Sure thing, I added that as a separate commit.

File: pkg/apis/crunchydata.com/v1/cluster.go [hunk 16]
```diff
@@ -100,7 +100,7 @@ type PgclusterSpec struct {
 	UserSecretName      string                   `json:"usersecretname"`
 	RootSecretName      string                   `json:"rootsecretname"`
 	PrimarySecretName   string                   `json:"primarysecretname"`
-	CollectSecretName   string                   `json:"collectSecretName"`
+	ExporterSecretName  string                   `json:"exporterSecretName"`
```
Comment (jkatz) [inline pkg/apis/crunchydata.com/v1/cluster.go:103]: @tjmoore4 Is this captured in the upgrade?

File: internal/operator/clusterutilities.go [hunk 17]
```diff
@@ -157,8 +158,8 @@ type DeploymentTemplateFields struct {
 	ContainerResources       string
 	NodeSelector             string
 	ConfVolume               string
-	CollectAddon             string
-	CollectVolume            string
+	ExporterAddon            string
+	ExporterVolume           string
```
Comment (jkatz) [inline internal/operator/clusterutilities.go:162]: I don't think this is used?

Comment (tjmoore4) [reply internal/operator/clusterutilities.go:166]: No, I missed that reference to 'ExporterVolume' during cleanup. It's removed now, thanks!

File: pkg/apis/crunchydata.com/v1/cluster.go (hunk 16 above)
Comment (tjmoore4) [reply pkg/apis/crunchydata.com/v1/cluster.go:103]: For the time being, this will be reverted to 'CollectSecretName' to support the current upgrade process. 
Hope to address this in the future.
